src/client/src/producer.rs

```diff
@@ -62,21 +101,86 @@ impl TopicProducer {
         V: Into<Vec<u8>>,
         I: IntoIterator<Item = (Option<K>, V)>,
     {
-        let replica = ReplicaKey::new(&self.topic, 0);
-        let spu_client = self.pool.create_serial_socket(&replica).await?;
-        debug!(
-            addr = spu_client.config().addr(),
-            "Connected to replica leader:"
-        );
-        let records: Vec<_> = records
+        let topics = self.pool.metadata.topics();
+        let topic_spec = topics
+            .lookup_by_key(&self.topic)
+            .await?
+            .ok_or_else(|| FluvioError::TopicNotFound(self.topic.to_string()))?
+            .spec;
+        let partition_count = topic_spec.partitions();
+        let partition_config = PartitionerConfig { partition_count };
+
+        let entries = records
             .into_iter()
-            .map(|(key, value): (Option<K>, V)| {
-                let key = key.map(|k| Bytes::from(k.into()));
-                let value = Bytes::from(value.into());
-                (key, value)
-            })
-            .collect();
-        send_records_raw(spu_client, &replica, records).await?;
+            .map::<(Option<Vec<u8>>, Vec<u8>), _>(|(k, v)| (k.map(|it| it.into()), v.into()));
+
+        // Calculate the partition for each entry
+        // Use a block scope to ensure we drop the partitioner lock
+        let iter = {
+            let mut iter = Vec::new();
+            let mut partitioner = self.partitioner.lock().await;
+            partitioner.update_config(partition_config);
+
+            for (k, v) in entries {
+                let partition = partitioner.partition(k.as_deref());
+                iter.push((partition, (k, v)));
+            }
+            iter
+        };
+
+        // Group all of the records by the partitions they belong to, then
+        // group all of the partitions by the SpuId that leads that partition
+        let mut map: HashMap<SpuId, HashMap<i32, Vec<DefaultRecord>>> = HashMap::new();
+        for (partition, (key, value)) in iter {
+            let replica_key = ReplicaKey::new(&self.topic, partition);
+            let partition_spec = self
+                .pool
+                .metadata
+                .partitions()
+                .lookup_by_key(&replica_key)
+                .await?
+                .ok_or_else(|| FluvioError::PartitionNotFound(self.topic.clone(), partition))?
+                .spec;
+            let leader = partition_spec.leader;
+            let key = key.map(|it| DefaultAsyncBuffer::new(it));
+            let record = DefaultAsyncBuffer::new(value);
+            let record = DefaultRecord::from((key, record));
+            map.entry(leader)
+                .or_insert_with(HashMap::new)
+                .entry(partition)
+                .or_insert_with(Vec::new)
+                .push(record);
```

Comment:
it is important to add unit test for this

Reply:
Added! I factored out the `group_by_spu` function and wrote a pretty good test for it, let me know what you think